app/tournament/tournament.js

```diff
@@ -0,0 +1,101 @@
+import * as player from "./player";
+import * as match from "./match";
+import * as utils from "../common/utils";
+
+export class Tournament {
+    constructor() {
+        this.players = [];
+        this.matches = [];
+    }
+
+    addPlayer(player) {
+        if (this.existsAlready(player)) {
+            throw new TournamentError(`Duplicate player: "${player.name}".`);
+        }
+        this.players.push(player);
+    }
+
+    existsAlready(player) {
+        for (var p of this.players) {
+            if (p.name.toLowerCase() === player.name.toLowerCase()) {
+                return true;
+            }
+        };
+        return false;
+    }
+
+    generate() {
+        this.checkNumberOfPlayers();
+        this.generateEmptyMatches();
+        this.generateTournamentBracket();
+        this.pairUpPlayers();
+    }
+
+    checkNumberOfPlayers() {
+        if (!this.enaughPlayers()) {
+            throw new TournamentError(`Too few players: ${this.players.length}`);
+        }
+        if (!this.hasProperNumberOfPlayers()){
+            this.fillUpWithGhosts();
+        }
+    }
+
+    enaughPlayers() {
+        return this.players.length >= 2;
+    }
+
+    hasProperNumberOfPlayers() {
```

Comment:
Do you want to unit test this logic? Move to a public utility method. Tournament doesn´t need to publish it.